### PR TITLE
ext: hal: st: move to modules

### LINKED ISSUE #16316: ST modules organization
## Introduction

Due to move of `ext/hal/st` to modules a reorganization of ST libraries is needed.
This issue is mainly a formal request about what was discussed in #16044 and summarized in https://github.com/zephyrproject-rtos/zephyr/pull/16044#issuecomment-493023435

### Problem description

`ext/hal/st` contains 2 different types of 'libraries' (in the most generic acceptance of the word):
- Code that is supposed to run on STM32 products (STM32Cube HAL/LL, Product specific IPC library). This restriction does not come from license type but purely that it is made to drive STM32 HW.
- Code that is made to drive ST products, with no restriction on the CPU on which it could run (could be STM32 or also HW from any other vendor). This is typically small libraries driving a small ST component (sensor, RF module, ..)

Following introduction of Zephyr module, all code under `ext/hal/st` should be moved under zephyr module. Moving it as a whole inside a single ST module would imply that anyone that want to use a ST component on a non ST product would have to pull the whole STM32Cube packages (> 220MB).

### Proposed change

Split ST libraries into 2 modules:
- `hal_stm32`: For the code **meant to run** on STM32
- `hal_st`: For the code **meant to drive** ST components 

## Detailed RFC

Move code under `ext/hal/st` in 2 ST specific modules, using the following trees:

```
hal_stm32
 └── stm32cube
 │   ├── common
 │   ├── stm32f1
 │   ├── ...
 │   └── stm32wb
 └── lib
     ├── stm32wb
     │   └── hci  (cf #14188)
     └── ...

hal_st
 └── stsensors
 │   ├── audio
 │   │   └── microphone
 │   ├─ ....
 │   ├── proximity
 │   │   └── vl53l0x
 │   └── stmemsc  (cf #14111)
 │       ├── a3g4250d
 │       ├── ais328dq
 │       └── ...
 └── strf
     ├── subg1 (cf #16086)
     └── ...
```

### Concerns and Unresolved Questions

As explained, these modules would not contain only HAL(s). 
It might make sense to use lib_st and lib_stm32 instead of hal_st and hal_stm32.
Unless 'hal' is made to be recognized as a zephyr keyword for this kind of stuff.

Second point would be feasibility of keeping git history following this migration. Keeping it would be great.

## Alternatives

Move `ext/hal/st` in a single ST module, with the drawbacks listed above.

### PR DESCRIPTION
**Pull request by nashif**
split ext/hal/st into two modules:

- hal_stm32: For the code meant to run on STM32
- hal_st: For the code meant to drive ST components

Fixes #16316

### CODE REVIEW

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Ack for the hal_stm32 part.
I'll let time to @avisconti to check the hal_st part.

**Comment by erwango** [comment]
@nashif , seems that hal_st leans under libs.
Is that because these are, in opposition to hal, not dedicated to a mcu family?

File: west.yml [hunk 1]
```diff
@@ -47,6 +47,12 @@ manifest:
     - name: hal_silabs
       revision: 9151e614c23997074acd1096a3e8a9e5c255d5b9
       path: modules/hal/silabs
+    - name: hal_st
+      revision: b6c9262eed68000e69c92bef6e820cdbd5b0a32b
+      path: modules/lib/st
```
Comment (avisconti) [inline west.yml]: Can we call it modules/hal/st ?
In here I'm planning to add also STdC HAL i/f to ST sensors.

**Review by avisconti — CHANGES_REQUESTED** [review]

**Review by avisconti — APPROVED** [review]
@erwango 
Do you agree in having 
modules/hal/st for hal_st
and
modules/hal/stm32 for hal_stm32 ?

**Comment by erwango** [comment]
@nashif , last though: can we change 'stm32' to 'stm32cube'? 
This will be more in line with our general naming.

**Comment by avisconti** [comment]
> @nashif , last though: can we change 'stm32' to 'stm32cube'?
> This will be more in line with our general naming.

agree!

**Comment by erwango** [comment]
> @nashif , last though: can we change 'stm32' to 'stm32cube'?

@nashif , @avisconti , actually coming back on that. stm32cube is already in hal_stm32 path (underneath), this is enough.
Let's keep as it is

**Pull request merged**